nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/avro/WriteAvroResultWithExternalSchema.java

```diff
@@ -78,14 +78,19 @@ protected void onBeginRecordSet() throws IOException {
     @Override
     public Map<String, String> writeRecord(final Record record) throws IOException {
         // If we are not writing an active record set, then we need to ensure that we write the
-        // schema information.
+        // schema information at the beginning and call flush at the end.
         if (!isActiveRecordSet()) {
             flush();
             schemaAccessWriter.writeHeader(recordSchema, getOutputStream());
         }
 
         final GenericRecord rec = AvroTypeUtil.createAvroRecord(record, avroSchema);
         datumWriter.write(rec, encoder);
+
+        if (!isActiveRecordSet()) {
+            flush();
```

Comment:
I did not really understand why `flush()` had been added in line 83, but I did not want to change the existing logic (though it could be removed with no risk I think).

I'd prefer `flush()` here because `write(RecordSet recordSet)` in `AbstractRecordSetWriter` handles its own `beginRecordSet()` / `finishRecordSet()` calls (and `flush()` is being called there). So I think `write(Record record)` should do the same.

Additionally, we can also call `flush()` in the `close()` again (just for sure to clean up the resource).

Adding 'flush()' only to 'close()' would not be the best solution in my opinion, because in that way this 'flush()' call was necessary for write(Record), but redundant for write(RecordSet), which is not consistent / straightforward.

Reply:
I think we can agree to disagree. But in general, most writer classes would leverage the buffered output stream to let it flush on full buffers and provide a configured/optimal writing strategy. Then flush the remaining data on close. If this is the case then all flushes outside the close are misplaced in my opinion. If the class utilizing the writer needs the data flushed it should be required to explicitly call flush(). If you look at the [csv counter part](https://github.com/apache/nifi/blob/rel/nifi-1.9.2/nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/csv/WriteCSVResult.java) it never flushes, and depends on the printers close method. Which could also be done here as well, by calling the close on the buffered outputsteam as well.

Going to stress this is mostly an opinion piece and mostly concerned about data loss, since it caused a lot of headache.

Reply:
Hi, thank you both for providing the fix! Please let me join the discussion :)

> I'd prefer flush() here because write(RecordSet recordSet) in AbstractRecordSetWriter handles its own beginRecordSet() / finishRecordSet() calls (and flush() is being called there). So I think write(Record record) should do the same.

@turcsanyip, probably I'm missing something, but I don't find any `flush()` call from AbstractRecordSetWriter `write(RecordSet recordSet)`, `beginRecordSet()` or `finishRecordSet()`. If no flush is called from there, the redundancy you worried about doesn't happen for the `write(RecordSet)` path even if we add `flush()` to `close()` at WriteAvroResultWithExternalSchema. Please correct me if I'm wrong on this.

Then, I'd prefer just adding `flush()` to `close()` as @pcgrenier proposed at #3568 

If both of you agree, I can merge #3573 first to have the unit tests, then #3568 to overwrite how to call flush(). Then commit history will look like this. How do you guys think?
https://github.com/ijokarumawak/nifi/commits/nifi-6419

Reply:
@ijokarumawak, sorry, my description was not clear. More precisely: `flush()` is called indirectly, `AbstractRecordSetWriter.finishRecordSet()` calls `onFinishRecordSet()` which is overridden in `WriteAvroResultWithExternalSchema` and calls `flush()`.

Generally speaking, I agree that calling `flush()` from client code is not a good practice and typically a code smell. My intention was only to keep consistency with the existing code.

Reply:
@turcsanyip I see, you're right, `flush()` is called indirectly from `onFinishRecordSet()`.

I don't think calling `flush()` from client code is a bad practice. Rather, I think it's a good practice as we developers don't know if each writer implementation actually flushes buffered data when its close is called. You may find these Stackoverflow threads interesting. Most people expect `flush()` is called automatically when they call `close()`, but relying on that assumption can cause issues like this, NIFI-6419.

In that sense, not calling `flush()` within `WriteAvroResultWithExternalSchema.close()` is correct. There's no contract that every close implementation should call flush first. However, since the RecordSetWriter/RecordWriter are used many places and we don't want to check each of them to confirm whether `flush()` is called properly, I think calling `flush()` at WriteAvroResultWithExternalSchema is a safer solution in this case. That's what both of you proposed.

We've been discussing where to add the `flush()` invocation. While I think both proposal make sense, I still prefer adding `flush()` at `close()`. Because, in order to fully utilize underlying buffer mechanism, writer implementation shouldn't call `flush()` too often. In this particular case, existing `writeRecord(Record record)` already calls `flush()`, so it may not be that important, though..

Let's consider following possible usage. A developer wants to use RecordWriter to write a custom file format which can have multiple Avro records separated by special character followed by some additional information. Then pseudo code would be:
```
try (output = some_file, writer = createWriter(output)) {
    while (hasMoreRecord()) {
        writer.write(record)
        output.write(additionalInfo)
    }
}
```

If we add `flush()` at `write()`, flush will be called at every iteration, and may result a unoptimized performance.

@pcgrenier already gave a thumbs-up above. If you agree by addressing NIFI-6419 by commits both of you in order to add flush at close, and helpful unit tests as I mentioned at the previous comment, I can merge them to master branch.
@turcsanyip How do you think? Is there any obvious reason we should add flush at `writeRecord()` instead?

Reply:
@ijokarumawak No, that will be fine with me.
I see the point in your example / pseudo code. For the same reason, the `flush()` at the beginning of `writeRecord()` should be removed too. Anyway, let's focus on the current bug and fix it only.

Reply:
Thanks @turcsanyip! Yeah, I thought the same thing. I cannot think of any necessity of the `flush()` at the beginning of `writeRecord()`, but let's focus on fixing the current issue this time.

I'm going to merge your commits. Thanks!